### PR TITLE
Implement Healthcheck status for keys loading

### PR DESCRIPTION
**Pull request by usmansaleem**
## PR Description
- Implement Healthcheck status for keys loading.
- Add AWS endpoint override cli option `--aws-endpoint-override` (useful for testing against localstack).

Sample healthcheck:
```
{
	"status": "UP",
	"checks": [
		{
			"id": "default-check",
			"status": "UP"
		},
		{
			"id": "keys-check",
			"status": "UP",
			"checks": [
				{
					"id": "aws-bulk-loading",
					"status": "UP",
					"data": {
						"keys-loaded": 2,
						"error-count": 0
					}
				},
				{
					"id": "config-files-loading",
					"status": "UP",
					"data": {
						"keys-loaded": 0,
						"error-count": 0
					}
				}
			]
		}
	],
	"outcome": "UP"
}
```

## Fixed Issue(s)

#715 
#730 
## Documentation

- [x] I thought about documentation and added the `doc-change-required` label to this PR if updates are required.

## Changelog

- [x] I thought about adding a changelog entry, and added one if I deemed necessary.

### CODE REVIEW

File: signing/src/test/java/tech/pegasys/web3signer/signing/config/SignerLoaderTest.java [hunk 1]
```diff
@@ -123,31 +127,35 @@ void wrongFileExtensionReturnsEmptySigner() throws IOException {
...
-    final List<ArtifactSigner> signerList =
-        Lists.newArrayList(new SignerLoader().load(missingDir, FILE_EXTENSION, signerParser));
+    final SecretValueResult<ArtifactSigner> result =
+        new SignerLoader().load(missingDir, FILE_EXTENSION, signerParser);
 
-    assertThat(signerList).isEmpty();
+    assertThat(result.getValues()).isEmpty();
+    // TODO: Should we return result with errorCount 1 if config directory is missing? Currently it
+    // is being treated as no error encountered
+    assertThat(result.getErrorCount()).isZero();
```
Comment (jframe) [inline signing/src/test/java/tech/pegasys/web3signer/signing/config/SignerLoaderTest.java:158]: I think your comment is correct, in the case where there is a directory error I would expect the an error to be registered

File: core/src/main/java/tech/pegasys/web3signer/core/Eth2Runner.java [hunk 2]
```diff
@@ -275,42 +281,109 @@ protected ArtifactSignerProvider createArtifactSignerProvider(
...
                             List.of(artifactSignerFactory),
                             YamlMapperFactory.createYamlMapper(
-                                config.getKeyStoreConfigFileMaxSize()))));
+                                config.getKeyStoreConfigFileMaxSize())));
+            super.registerHealthCheckProcedure(
+                KEYS_CHECK_CONFIG_FILE_LOADING,
+                promise -> {
+                  final JsonObject statusJson =
+                      new JsonObject()
+                          .put("keys-loaded", results.getValues().size())
```
[the remark this answers is not in the document]
Comment (usmansaleem) [reply core/src/main/java/tech/pegasys/web3signer/core/Eth2Runner.java:298]: correct, these will be re-evaluated during reload.

File: core/src/main/java/tech/pegasys/web3signer/core/Eth2Runner.java [hunk 3]
```diff
@@ -275,42 +281,109 @@ protected ArtifactSignerProvider createArtifactSignerProvider(
...
+                promise -> {
+                  final JsonObject statusJson =
+                      new JsonObject()
+                          .put("keys-loaded", keystoreSignersResult.getValues().size())
+                          .put("error-count", keystoreSignersResult.getErrorCount());
+                  promise.complete(
+                      keystoreSignersResult.getErrorCount() > 0
+                          ? Status.KO(statusJson)
+                          : Status.OK(statusJson));
+                });
```
[the remark this answers is not in the document]
Comment (usmansaleem) [reply core/src/main/java/tech/pegasys/web3signer/core/Eth2Runner.java]: done.

File: signing/src/test/java/tech/pegasys/web3signer/signing/config/SignerLoaderTest.java (hunk 1 above)
Comment (usmansaleem) [reply signing/src/test/java/tech/pegasys/web3signer/signing/config/SignerLoaderTest.java]: Updated logic ... if config dir can't be read, the error count will not be zero.

File: acceptance-tests/src/test/java/tech/pegasys/web3signer/tests/bulkloading/AzureKeyVaultAcceptanceTest.java [hunk 4]
```diff
@@ -72,6 +73,13 @@ void invalidVaultParametersFailsToLoadKeys() {
...
     final Response response = signer.callApiPublicKeys(KeyType.BLS);
     response.then().statusCode(200).contentType(ContentType.JSON).body("", hasSize(0));
+
+    signer
+        .healthcheck()
+        .then()
+        .statusCode(503)
+        .contentType(ContentType.JSON)
+        .body("status", equalTo("DOWN"));
   }
```
[the remark this answers is not in the document]
Comment (usmansaleem) [reply acceptance-tests/src/test/java/tech/pegasys/web3signer/tests/bulkloading/AzureKeyVaultAcceptanceTest.java:83]: done for Azure as well.

File: signing/src/main/java/tech/pegasys/web3signer/signing/BlsKeystoreBulkLoader.java [hunk 5]
```diff
@@ -39,36 +39,66 @@
...
+    final List<SecretValueResult<ArtifactSigner>> resultList =
+        keystoreFiles.parallelStream()
+            .map(
+                keystoreFile ->
+                    createSignerForKeystore(
+                        keystoreFile,
+                        key -> Files.readString(passwordsDirectory.resolve(key + ".txt"))))
+            .collect(Collectors.toList());
+
+    return combineResults(resultList);
```
[the remark this answers is not in the document]
Comment (usmansaleem) [reply signing/src/main/java/tech/pegasys/web3signer/signing/BlsKeystoreBulkLoader.java]: done, using reduce now.

File: signing/src/main/java/tech/pegasys/web3signer/signing/BlsKeystoreBulkLoader.java [hunk 6]
```diff
@@ -39,36 +39,59 @@
 public class BlsKeystoreBulkLoader {
   private static final Logger LOG = LogManager.getLogger();
 
-  public Collection<ArtifactSigner> loadKeystoresUsingPasswordDir(
+  public SecretValueResult<ArtifactSigner> loadKeystoresUsingPasswordDir(
```
Comment (jframe) [inline signing/src/main/java/tech/pegasys/web3signer/signing/BlsKeystoreBulkLoader.java:42]: This doesn't seem like the right type to expose, we are not returning the secret values but artifacts. Can we create a type to hold the collection of artifactSigners and the number of errors.

File: signing/src/main/java/tech/pegasys/web3signer/signing/BlsKeystoreBulkLoader.java [hunk 7]
```diff
@@ -39,36 +39,59 @@
...
             keystoreFile ->
                 createSignerForKeystore(
                     keystoreFile,
                     key -> Files.readString(passwordsDirectory.resolve(key + ".txt"))))
-        .flatMap(Optional::stream)
-        .collect(Collectors.toList());
+        .reduce(
+            SecretValueResult.newInstance(new HashSet<>(), 0),
+            (r1, r2) -> {
+              r1.merge(r2);
```
Comment (jframe) [inline signing/src/main/java/tech/pegasys/web3signer/signing/BlsKeystoreBulkLoader.java:61]: nit: if this was a static function that returned a SecretValueResult then you use a method reference here

File: signing/src/main/java/tech/pegasys/web3signer/signing/BlsKeystoreBulkLoader.java (hunk 6 above)
Comment (usmansaleem) [reply signing/src/main/java/tech/pegasys/web3signer/signing/BlsKeystoreBulkLoader.java:42]: But this is the same type which is used by Aws bulkloading and Azure bulkloading as well.

File: core/src/main/java/tech/pegasys/web3signer/core/Eth2Runner.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (jframe) [reply core/src/main/java/tech/pegasys/web3signer/core/Eth2Runner.java]: Can a test be added for the reload case, I think have a test for reloading config so shouldn't be too tricky hopefully 🤞

File: acceptance-tests/src/test/java/tech/pegasys/web3signer/tests/bulkloading/AzureKeyVaultAcceptanceTest.java [hunk 8]
```diff
@@ -58,6 +59,13 @@ void ensureSecretsInKeyVaultAreLoadedAndReportedViaPublicKeysApi() {
 
     final Response response = signer.callApiPublicKeys(KeyType.BLS);
     response.then().statusCode(200).contentType(ContentType.JSON).body("", contains(EXPECTED_KEY));
+
+    signer
+        .healthcheck()
+        .then()
+        .statusCode(200)
+        .contentType(ContentType.JSON)
+        .body("status", equalTo("UP"));
```
[the remark this answers is not in the document]
Comment (usmansaleem) [reply acceptance-tests/src/test/java/tech/pegasys/web3signer/tests/bulkloading/AzureKeyVaultAcceptanceTest.java:68]: Added it in AWS Secrets Manager AT.

File: acceptance-tests/src/test/java/tech/pegasys/web3signer/tests/signing/BlsSigningAcceptanceTest.java [hunk 9]
```diff
@@ -178,10 +179,16 @@ public void ableToSignUsingAws() throws JsonProcessingException {
...
+    final Optional<URI> awsEndpointOverride =
+        System.getenv("AWS_ENDPOINT_OVERRIDE") != null
+            ? Optional.of(URI.create(System.getenv("AWS_ENDPOINT_OVERRIDE")))
+            : Optional.empty();
     final String publicKey = KEY_PAIR.getPublicKey().toString();
 
     final AwsSecretsManagerUtil awsSecretsManagerUtil =
-        new AwsSecretsManagerUtil(region, rwAwsAccessKeyId, rwAwsSecretAccessKey);
+        new AwsSecretsManagerUtil(
+            region, rwAwsAccessKeyId, rwAwsSecretAccessKey, awsEndpointOverride);
```
[the remark this answers is not in the document]
Comment (usmansaleem) [reply acceptance-tests/src/test/java/tech/pegasys/web3signer/tests/signing/BlsSigningAcceptanceTest.java:191]: Added healthcheck AT related to config-files-loading.

**Review by jframe — APPROVED** [review]

**Pull request merged**